typing/types.ml

```diff
@@ -996,19 +1019,135 @@ end
+   saying but it also needs to expose a total order.
+
+   Someday, it's probably desirable to merge this, and make it compatible, with
+   [Ctype.eq_type], though that seems quite hard.
+*)
+let rec best_effort_compare_type_expr te1 te2 =
+  if te1 == te2 || repr te1 == repr te2 then 0
+  else
+    let rank ty =
+      match ty.desc with
+      (* Types which must be compared by id *)
+      | Tvar _
+      | Tobject (_, _)
+      | Tfield (_, _, _, _)
+      | Tnil
+      | Tlink _
+      | Tsubst (_, _)
+      | Tvariant _
+      | Tpackage (_, _)
+      | Tarrow (_, _, _, _)
+        -> -ty.id
+      (* Types which we know how to compare structurally*)
+      | Ttuple _ -> 2
+      | Tunboxed_tuple _ -> 3
+      | Tunivar _ -> 4
+      | Tconstr (_, _, _) -> 5
+      | Tpoly (_, _) -> 6
+    in
+    match te1.desc, te2.desc with
+    | Ttuple elts1, Ttuple elts2
+    | Tunboxed_tuple elts1, Tunboxed_tuple elts2 ->
+      List.compare
+        (fun (l1, te1) (l2, te2) ->
+           let l = Option.compare String.compare l1 l2 in
+           if l = 0 then best_effort_compare_type_expr te1 te2 else l
+        )
+        elts1
+        elts2
+    | Tconstr (p1, args1, _), Tconstr (p2, args2, _) ->
+      let p = Path.compare p1 p2 in
+      if p = 0
+      then List.compare best_effort_compare_type_expr args1 args2
+      else p
+    | Tpoly (t1, ts1), Tpoly (t2, ts2) ->
+      List.compare best_effort_compare_type_expr (t1 :: ts1) (t2 :: ts2)
+    | _, _ -> rank te1 - rank te2
+
+let () =
+  With_bounds_type_map_unsafe.compare_type_expr :=
+    (fun obj1 obj2 ->
+       best_effort_compare_type_expr
+         (Obj.obj obj1 : type_expr)
+         (Obj.obj obj2 : type_expr))
+
+module With_bounds_types : sig
+  (* Note that only the initially needed bits of [Stdlib.Map.S] are exposed here; feel
+     free to expose more functions if you need them! *)
+  type info := With_bounds_type_info.t
+  type t := with_bounds_types
+
+  val empty : t
+  val is_empty : t -> bool
+  val to_seq : t -> (type_expr * info) Seq.t
+  val of_list : (type_expr * info) list -> t
+  val update : type_expr -> (info option -> info option) -> t -> t
+  val merge
+    : (type_expr -> info option -> info option -> info option) ->
+    t -> t -> t
+  val map : (info -> info) -> t -> t
+
+  module Non_empty : sig
+    type maybe_empty := t
+    type t = nonempty_with_bounds_types
+
+    val of_maybe_empty : maybe_empty -> t option
+    val to_maybe_empty : t -> maybe_empty
+    val singleton : type_expr -> info -> t
+    val update : type_expr -> (info option -> info option) -> t -> t
+    val merge
+      : (type_expr -> info option -> info option -> info option) ->
+      t -> t -> t
+    val to_seq : t -> (type_expr * info) Seq.t
+    val map : (info -> info) -> t -> t
+  end
+end = struct
+  module U = With_bounds_type_map_unsafe
+  include U
+
+  let to_seq t = to_seq t |> Seq.map (fun (ty, ti) -> ((Obj.obj ty : type_expr), ti))
```

Comment:
I would abstract out the call to `Obj.obj` into a function with a type signature. I see it's only used twice, but it would be very easy to make a disastrous mistake around this.